torchqdynamics/utils/td_tensor.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import get_args
+
+import torch
+from torch import Tensor
+
+from .tensor_types import OperatorLike, TDOperatorLike, cdtype, rdtype, to_tensor
+
+
+def to_tdtensor(
+    x: TDOperatorLike,
+    dtype: torch.dtype | None = None,
+    device: torch.device | None = None,
+    is_complex: bool = False,
+) -> TDTensor:
+    """Convert a `TDOperatorLike` object to a `TDTensor` object."""
+    if is_complex:
+        dtype = cdtype(dtype)
+
+    if isinstance(x, get_args(OperatorLike)):
+        x = to_tensor(x, dtype=dtype, device=device, is_complex=is_complex)
+        return ConstantTDTensor(x, dtype=dtype, device=device)
+    elif callable(x):
+        if dtype is None and not is_complex:
+            dtype = rdtype(dtype)
+        if device is None:
+            device = torch.device('cpu')
+        return CallableTDTensor(x, dtype=dtype, device=device)
+
+
+class TDTensor(ABC):
+    @abstractmethod
+    def __call__(self, t: float) -> Tensor:
+        """Evaluate a `TDTensor` at a given time"""
+        pass
+
+    @abstractmethod
+    def size(self, dim: int) -> int:
+        """Size of a `TDTensor` along a given dimension."""
+        pass
+
+    @abstractmethod
+    def dim(self) -> int:
+        """Get the number of dimensions of a `TDTensor`."""
+        pass
+
+    @abstractmethod
+    def unsqueeze(self, dim: int) -> TDTensor:
+        """Unsqueeze a `TDTensor` at position `dim`."""
+        pass
+
+    @abstractmethod
+    def has_changed(self, t: float) -> bool:
+        """Checks whether the `TDTensor` has changed since the last call."""
+        pass
+
+
+class ConstantTDTensor(TDTensor):
+    def __init__(
+        self, tensor: Tensor, *, dtype: torch.dtype, device: torch.device | None
+    ):
+        self._tensor = tensor
+        self.dtype = dtype or tensor.dtype
+        self.device = device or tensor.device
```

Comment:
```suggestion
    def __init__(self, tensor: Tensor):
        self._tensor = tensor
```
The `dtype` and `device` we need is the one of `tensor`, the conversion already happened with the call to `to_tensor` within `to_tdtensor`.

Reply:
In the end I moved the `check_callable` function outside of the class. I think it's better this way, it makes the class lighter which should be favorable for performance.